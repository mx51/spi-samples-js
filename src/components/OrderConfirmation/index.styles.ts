import { alpha, createStyles, makeStyles, Theme } from '@material-ui/core';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      height: `calc(100vh - 120px)`,
      position: 'sticky',
      top: theme.spacing(6),
      zIndex: 1020,
    },
    gridStyles: {
      margin: theme.spacing(3),
    },
    payLabel: {
      fontSize: '3rem',
      fontWeight: 300,
      lineHeight: '4.08rem',
      paddingRight: '3rem',
    },
    paper: {
      border: 'none',
      paddingRight: '1rem',
      alignItems: 'center',
      textAlign: 'center',
    },
    orderTotalBtn: {
      '&:hover': {
        background: 'none',
      },
    },
    radioGroup: {
      flexGrow: 1,
      overflowY: 'auto',
    },
    orderTotalInputField: {
      marginRight: theme.spacing(2),
      padding: '2rem',
      letterSpacing: '0.02rem',
      fontSize: '3rem',
      fontWeight: 'normal',
    },
    label: {
      marginTop: theme.spacing(2),
      marginBottom: theme.spacing(0.5),
    },
    paymentTypeBtnLabel: {
      fontWeight: 500,
      fontSize: '1.5rem',
      lineHeight: '1.5rem',
    },

    paymentTypeBtn: {
      padding: theme.spacing(2),
      backgroundColor: theme.palette.error.dark,
      '&:hover, &:active, &:focus': {
        background: alpha(theme.palette.error.dark, 0.75),
        color: theme.palette.common.white,
      },

      width: '100%',
      margin: theme.spacing(1),
    },
    radioBtn: {
      color: theme.palette.primary.main,
    },
    keypadDrawerPaper: {
      width: '33.34%',
      backgroundColor: 'transparent',
      border: 0,
    },
  })
);
export default useStyles;
