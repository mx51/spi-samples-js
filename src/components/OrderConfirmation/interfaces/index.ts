export interface IOrderConfirmation {
  title: string;
  pathname: string;
  editSubtotal: boolean;
}

export interface ITitleStrategy {
  Pay: string;
  Cashout: string;
  Refund: string;
}

export interface IProps {
  setShowTransactionProgressModal: (show: boolean) => void;
}
