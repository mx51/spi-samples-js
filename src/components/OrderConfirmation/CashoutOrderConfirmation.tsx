import React from 'react';
import { useSelector } from 'react-redux';
import { Box, Button } from '@material-ui/core';
import { IProps } from './interfaces';
import useStyles from './index.styles';
import { initiateCashoutOnlyTx } from '../../utils/common/purchase/purchaseHelper';
import {
  orderCashoutAmountSelector,
  orderSurchargeAmountSelector,
} from '../../redux/reducers/ProductSlice/productSelector';

export const CashoutOrderConfirmation: React.FC<IProps> = ({ setShowTransactionProgressModal, transactionHandler }) => {
  const classes = useStyles();
  const surchargeAmount: number = useSelector(orderSurchargeAmountSelector);
  const cashoutAmount: number = useSelector(orderCashoutAmountSelector);

  return (
    <Box display="flex" justifyContent="space-evenly">
      <Button
        variant="contained"
        color="primary"
        size="large"
        disabled={cashoutAmount <= 0 || !transactionHandler}
        focusRipple
        classes={{ root: classes.paymentTypeBtn, label: classes.paymentTypeBtnLabel }}
        onClick={() => {
          setShowTransactionProgressModal(true);
          initiateCashoutOnlyTx(transactionHandler!.terminalId!, cashoutAmount, surchargeAmount);
        }}
      >
        Cashout
      </Button>
    </Box>
  );
};
