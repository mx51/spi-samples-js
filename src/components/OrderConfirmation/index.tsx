import React, { useState } from 'react';
import {
  Box,
  Button,
  Divider,
  Drawer,
  Grid,
  Icon,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Paper,
  Radio,
  RadioGroup,
  Typography,
} from '@material-ui/core';

import CreateIcon from '@material-ui/icons/Create';
import { useDispatch, useSelector } from 'react-redux';
import {
  orderCashoutAmountSelector,
  orderSurchargeAmountSelector,
  orderTipAmountSelector,
  orderTotalSelector,
} from '../../redux/reducers/ProductSlice/productSelector';
import { clearAllProducts } from '../../redux/reducers/ProductSlice/productSlice';
import { updateSelectedTerminal } from '../../redux/reducers/SelectedTerminalSlice/selectedTerminalSlice';
import selectedTerminalIdSelector from '../../redux/reducers/SelectedTerminalSlice/selectedTerminalSliceSelector';
import { ITerminalProps } from '../../redux/reducers/TerminalSlice/interfaces';
import {
  pairedConnectedTerminalList,
  terminalInstance,
} from '../../redux/reducers/TerminalSlice/terminalsSliceSelectors';
import currencyFormat from '../../utils/common/intl/currencyFormatter';
import { initiatePurchase, initiateMotoPurchase } from '../../utils/common/purchase/purchaseHelper';
import KeyPad from '../KeyPad';
import TransactionProgressModal from '../TransactionProgressModal';
import useStyles from './index.styles';

function PayNow(): React.ReactElement {
  const dispatch = useDispatch();

  const classes = useStyles();
  const originalTotalAmount: number = useSelector(orderTotalSelector);
  const surchargeAmount: number = useSelector(orderSurchargeAmountSelector);
  const tipAmount: number = useSelector(orderTipAmountSelector);
  const cashoutAmount: number = useSelector(orderCashoutAmountSelector);
  const terminals = useSelector(pairedConnectedTerminalList);
  const selectedTerminal = useSelector(selectedTerminalIdSelector);
  const currentTerminal = useSelector(terminalInstance(selectedTerminal)) as ITerminalProps;

  const clearAllProductsAction = () => {
    dispatch(clearAllProducts());
  };

  const [displayKeypad, setDisplayKeypad] = useState<boolean>(false);
  const [totalAmount, setTotalAmount] = useState<number>(originalTotalAmount);
  const [showTransactionProgressModal, setShowTransactionProgressModal] = useState<boolean>(false);

  function selectTerminal(terminalId: string) {
    dispatch(updateSelectedTerminal(terminalId));
  }

  return (
    <>
      <Drawer
        anchor="right"
        open={displayKeypad}
        classes={{
          paper: classes.keypadDrawerPaper,
        }}
      >
        <KeyPad
          open={displayKeypad}
          title="Override Purchase amount"
          subtitle="Enter purchase amount"
          defaultAmount={totalAmount}
          onClose={() => {
            setDisplayKeypad(false);
          }}
          onAmountChange={(amount) => {
            setTotalAmount(amount);
            setDisplayKeypad(false);
            clearAllProductsAction();
          }}
        />
      </Drawer>
      <Grid container justifyContent="center">
        <Grid item xs={7} className={classes.gridStyles}>
          <Box className={classes.root} display="flex" flexDirection="column">
            <Box display="flex" alignItems="center" justifyContent="space-between">
              <Box flex="1">
                <Typography className={classes.payLabel}>Pay</Typography>
              </Box>
              <Button className={classes.orderTotalBtn} onClick={() => setDisplayKeypad(true)}>
                <Box
                  flex="1"
                  display="flex"
                  className={classes.paper}
                  onClick={() => setDisplayKeypad(true)}
                  component={Paper}
                >
                  <Box className={classes.orderTotalInputField} flex="1">
                    {currencyFormat(totalAmount / 100)}
                  </Box>
                  <Box>
                    <Icon>
                      <CreateIcon />
                    </Icon>
                  </Box>
                </Box>
              </Button>
            </Box>
            <Typography className={classes.label}>Select terminal</Typography>
            <Divider />
            <RadioGroup className={classes.radioGroup} aria-label="terminalList" name="terminalList">
              <Box>
                <List>
                  {terminals.map((terminal) => (
                    <ListItem key={terminal.id} dense disableGutters>
                      <ListItemIcon>
                        <Radio
                          onClick={() => selectTerminal(terminal.id)}
                          className={classes.radioBtn}
                          checked={terminal.id === selectedTerminal}
                          value={terminal.id}
                          name="terminal"
                        />
                      </ListItemIcon>
                      <ListItemText
                        primary={terminal.posId}
                        secondary={`${terminal.deviceAddress} S/N ${terminal.serialNumber}`}
                      />
                    </ListItem>
                  ))}
                </List>
              </Box>
            </RadioGroup>
            <Typography className={classes.label}>Select payment method</Typography>
            <Divider />
            {showTransactionProgressModal && (
              <TransactionProgressModal
                transactionType={currentTerminal?.txFlow?.type ?? ''}
                isFinished={currentTerminal?.txFlow?.finished ?? false}
                isSuccess={currentTerminal?.txFlow?.success === 'Success'}
                onCancelTransaction={() => {
                  setShowTransactionProgressModal(false);
                }}
                onDone={() => {
                  setShowTransactionProgressModal(false);
                }}
              />
            )}
            <Box display="flex" justifyContent="space-evenly">
              <Button
                variant="contained"
                color="primary"
                size="large"
                disabled={!selectedTerminal}
                focusRipple
                classes={{ root: classes.paymentTypeBtn, label: classes.paymentTypeBtnLabel }}
                onClick={() => {
                  setShowTransactionProgressModal(true);
                  initiatePurchase(selectedTerminal, totalAmount, tipAmount, cashoutAmount, surchargeAmount, false);
                }}
              >
                Card
              </Button>
              <Button
                variant="contained"
                color="primary"
                size="large"
                disabled={!selectedTerminal}
                focusRipple
                classes={{ root: classes.paymentTypeBtn, label: classes.paymentTypeBtnLabel }}
                onClick={() => {
                  setShowTransactionProgressModal(true);
                  initiateMotoPurchase(selectedTerminal, totalAmount, surchargeAmount);
                }}
              >
                Moto
              </Button>
            </Box>
          </Box>
        </Grid>
      </Grid>
    </>
  );
}

export default PayNow;
