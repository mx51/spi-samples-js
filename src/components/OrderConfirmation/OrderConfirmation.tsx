import React, { useState } from 'react';
import {
  Box,
  Button,
  Divider,
  Drawer,
  Grid,
  Icon,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Paper,
  Radio,
  RadioGroup,
  Typography,
} from '@material-ui/core';
import CreateIcon from '@material-ui/icons/Create';
import { useDispatch, useSelector } from 'react-redux';
import {
  PATH_CASH_OUT,
  PATH_REFUND,
  PATH_PRE_AUTH,
  TEXT_PURCHASE,
  TEXT_REFUND,
  TEXT_CASHOUT,
  PATH_PAY_NOW,
} from '../../definitions/constants/routerConfigs';
import { TxFlowState } from '../../definitions/constants/terminalConfigs';
import { orderTotalSelector } from '../../redux/reducers/ProductSlice/productSelector';
import { addCashoutAmount, addRefundAmount, clearProductsOnly } from '../../redux/reducers/ProductSlice/productSlice';
import { updateSelectedTerminal } from '../../redux/reducers/SelectedTerminalSlice/selectedTerminalSlice';
import selectedTerminalIdSelector from '../../redux/reducers/SelectedTerminalSlice/selectedTerminalSliceSelector';
import { ITerminalProps } from '../../redux/reducers/TerminalSlice/interfaces';
import { updateTxFlow } from '../../redux/reducers/TerminalSlice/terminalsSlice';
import {
  pairedConnectedTerminalList,
  terminalInstance,
  terminalTxFlowFinishedTracker,
  terminalTxFlowReceipt,
} from '../../redux/reducers/TerminalSlice/terminalsSliceSelectors';
import currencyFormat from '../../utils/common/intl/currencyFormatter';
import { cancelTransaction, setTerminalToIdle } from '../../utils/common/purchase/purchaseHelper';
import KeyPad from '../KeyPad';
import TransactionProgressModal from '../TransactionProgressModal';
import UnknownTransactionModal from '../UnknownTransactionModal';
import useStyles from './index.styles';
import { IOrderConfirmation, ITitleStrategy } from './interfaces';
import { CashoutOrderConfirmation } from './CashoutOrderConfirmation';
import { PayNowOrderConfirmation } from './PayNowOrderConfirmation';
import { PreAuthOrderConfirmation } from './PreAuthOrderConfirmation';
import { RefundOrderConfirmation } from './RefundOrderConfirmation';
import { updatePreAuthParams } from '../../redux/reducers/PreAuth/preAuthSlice';
import { preAuthSelector } from '../../redux/reducers/PreAuth/preAuthSelector';

type ComponentByPathNameKeys = typeof PATH_PRE_AUTH | typeof PATH_REFUND | typeof PATH_PAY_NOW | typeof PATH_CASH_OUT;

function OrderConfirmation({ title, pathname, editSubtotal }: IOrderConfirmation): React.ReactElement {
  const cashoutPage = pathname === PATH_CASH_OUT;
  const refundPage = pathname === PATH_REFUND;
  const preAuthPage = pathname === PATH_PRE_AUTH;

  const dispatch = useDispatch();
  const classes = useStyles();
  const terminals = useSelector(pairedConnectedTerminalList);
  const selectedTerminal = useSelector(selectedTerminalIdSelector);
  const currentTerminal = useSelector(terminalInstance(selectedTerminal)) as ITerminalProps | undefined;
  const isFinished = useSelector(terminalTxFlowFinishedTracker(selectedTerminal)) ?? false;
  const receipt = useSelector(terminalTxFlowReceipt(selectedTerminal));
  const successStatus = currentTerminal?.txFlow?.success;
  const isUnknownState = isFinished && successStatus === TxFlowState.Unknown;
  const [displayKeypad, setDisplayKeypad] = useState<boolean>(false);
  const [showTransactionProgressModal, setShowTransactionProgressModal] = useState<boolean>(false);
  const [toShowUnknownTransaction, setToShowUnknownTransaction] = useState<boolean>(false);
  const [showUnknownTransactionModal, setShowUnknownTransactionModal] = useState<boolean>(true);
  const totalAmount = useSelector(orderTotalSelector);
  const preAuth = useSelector(preAuthSelector);

  const clearProductsOnlyAction = () => {
    dispatch(clearProductsOnly());
  };

  function selectTerminal(terminalId: string) {
    dispatch(updateSelectedTerminal(terminalId));
  }

  function updateUnknownTerminalState(success: string) {
    setTerminalToIdle(selectedTerminal);
    setShowUnknownTransactionModal(false);
    if (currentTerminal?.txFlow != null) {
      dispatch(
        updateTxFlow({
          id: selectedTerminal,
          txFlow: { ...currentTerminal?.txFlow, finished: true, success },
        })
      );
    }
  }

  const titleStrategy: ITitleStrategy = {
    Pay: `Override ${TEXT_PURCHASE}`,
    [TEXT_CASHOUT]: TEXT_CASHOUT,
    [TEXT_REFUND]: TEXT_REFUND,
  };

  function getTitleForKeypad(): string {
    return title in titleStrategy ? (titleStrategy as unknown as Record<string, keyof ITitleStrategy>)[title] : title;
  }

  const componentByPathName: Record<ComponentByPathNameKeys, JSX.Element> = {
    '/pre-auth': <PreAuthOrderConfirmation setShowTransactionProgressModal={setShowTransactionProgressModal} />,
    '/refund': <RefundOrderConfirmation setShowTransactionProgressModal={setShowTransactionProgressModal} />,
    '/paynow': <PayNowOrderConfirmation setShowTransactionProgressModal={setShowTransactionProgressModal} />,
    '/cashout': <CashoutOrderConfirmation setShowTransactionProgressModal={setShowTransactionProgressModal} />,
  };

  const renderOrderConfirmationByAction = () => componentByPathName[pathname as ComponentByPathNameKeys];

  return (
    <>
      <Drawer
        anchor="right"
        open={displayKeypad}
        classes={{
          paper: classes.keypadDrawerPaper,
        }}
      >
        <KeyPad
          open={displayKeypad}
          title={getTitleForKeypad()}
          subtitle={`Enter ${getTitleForKeypad().toLowerCase()} amount`}
          defaultAmount={totalAmount}
          onClose={() => {
            setDisplayKeypad(false);
          }}
          onAmountChange={(amount) => {
            setDisplayKeypad(false);

            if (cashoutPage) {
              dispatch(addCashoutAmount(amount));
            } else if (refundPage) {
              dispatch(addRefundAmount(amount));
            } else if (preAuthPage) {
              dispatch(
                updatePreAuthParams({
                  key: 'UPDATE_CURRENT_AMOUNT',
                  value: amount,
                })
              );
            }
            clearProductsOnlyAction();
          }}
        />
      </Drawer>
      <Grid container justifyContent="center">
        <Grid item xs={7} className={classes.gridStyles}>
          <Box className={classes.root} display="flex" flexDirection="column">
            <Box display="flex" alignItems="center" justifyContent="space-between">
              <Box flex="1">
                <Typography className={classes.payLabel}>{title}</Typography>
              </Box>
              <Button
                data-testid="orderTotalButton"
                className={classes.orderTotalBtn}
                disableRipple={!editSubtotal}
                disabled={!editSubtotal}
                onClick={() => setDisplayKeypad(true)}
              >
                <Box flex="1" display="flex" className={classes.paper} component={Paper}>
                  <Box className={classes.orderTotalInputField} flex="1">
                    {preAuthPage ? currencyFormat(preAuth.currentAmount / 100) : currencyFormat(totalAmount / 100)}
                  </Box>
                  {editSubtotal && (
                    <Box>
                      <Icon>
                        <CreateIcon />
                      </Icon>
                    </Box>
                  )}
                </Box>
              </Button>
            </Box>
            <Typography className={classes.label}>Select terminal</Typography>
            <Divider />
            <RadioGroup className={classes.radioGroup} aria-label="terminalList" name="terminalList">
              <Box>
                <List>
                  {terminals.map((terminal) => (
                    <ListItem key={terminal.id} dense disableGutters onClick={() => selectTerminal(terminal.id)}>
                      <ListItemIcon>
                        <Radio
                          className={classes.radioBtn}
                          checked={terminal.id === selectedTerminal}
                          value={terminal.id}
                          name="terminal"
                        />
                      </ListItemIcon>
                      <ListItemText
                        primary={terminal.posId}
                        secondary={`${terminal.deviceAddress} S/N ${terminal.serialNumber}`}
                      />
                    </ListItem>
                  ))}
                </List>
              </Box>
            </RadioGroup>

            {renderOrderConfirmationByAction()}

            {showUnknownTransactionModal && isUnknownState && (
              <UnknownTransactionModal
                onSuccessTransaction={() => {
                  updateUnknownTerminalState('Success');
                  setToShowUnknownTransaction(true);
                }}
                onFailedTransaction={() => {
                  updateUnknownTerminalState('Failed');
                  setToShowUnknownTransaction(true);
                }}
              />
            )}
            {showTransactionProgressModal && (!isUnknownState || toShowUnknownTransaction) && (
              <TransactionProgressModal
                terminalId={selectedTerminal}
                transactionType={currentTerminal?.txFlow?.type ?? ''}
                transactionDesc={receipt?.hostResponseText ?? ''}
                isFinished={isFinished}
                isSuccess={successStatus === 'Success'}
                onCancelTransaction={() => {
                  cancelTransaction(selectedTerminal);
                }}
                onRetryTransaction={() => {
                  setShowTransactionProgressModal(false);
                }}
                onDone={() => {
                  setShowTransactionProgressModal(false);
                }}
              />
            )}
          </Box>
        </Grid>
      </Grid>
    </>
  );
}

export default OrderConfirmation;
