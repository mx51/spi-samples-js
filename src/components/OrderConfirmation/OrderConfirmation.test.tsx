import React from 'react';
import { fireEvent, screen, cleanup, waitFor } from '@testing-library/react';
import * as redux from 'react-redux';
import OrderConfirmation from '.';
import { PATH_CASH_OUT, PATH_PAY_NOW, PATH_REFUND } from '../../definitions/constants/routerConfigs';
import { TxFlowState } from '../../definitions/constants/terminalConfigs';
import { store } from '../../redux/store';
import spiService from '../../services/spiService';
import mockWithRedux, {
  defaultMockPairFormParams,
  mockReceiptResponse,
  mockRefundTxFlow,
  mockTerminalInstanceId,
  pairedMockTerminals,
} from '../../utils/tests/common';

describe('Test <OrderConfirmation />', () => {
  let dispatch: Any;
  let customStore: Any;

  beforeEach(() => {
    dispatch = jest.fn();
    const spyDispatch = jest.spyOn(redux, 'useDispatch');
    spyDispatch.mockReturnValue(dispatch);
    customStore = {
      ...store,
      getState: () => ({
        common: { showFlowPanel: false, acquireConfirmPairingFlow: false },
        pairForm: defaultMockPairFormParams,
        terminals: pairedMockTerminals,
        products: {
          keypadAmount: 0,
          surchargeAmount: 100,
          tipAmount: 0,
          cashoutAmount: 1000,
          products: [],
        },
        selectedTerminal: { selectedTerminalId: mockTerminalInstanceId },
      }),
    };
  });

  afterEach(cleanup);

  test('should match OrderConfirmationPage snapshot test', () => {
    // Arrange
    const container = mockWithRedux(<OrderConfirmation title="title" pathname="/" currentAmount={500} />, customStore);

    // Assert
    expect(container).toMatchSnapshot();
  });

  test('should clearAllProducts on amount override ', () => {
    // Arrange
    mockWithRedux(<OrderConfirmation title="title" pathname="/" currentAmount={500} />, customStore);
    fireEvent.click(screen.getByText(/test/i));

    // Assert
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0][0]).toEqual({
      payload: '123-123-123',
      type: 'selectedTerminal/updateSelectedTerminal',
    });
  });

  test('should select a terminal from list', () => {
    // Arrange
    mockWithRedux(<OrderConfirmation title="title" pathname="/" currentAmount={500} />, customStore);
    fireEvent.click(screen.getByText(/test/i));

    // Assert
    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0][0]).toEqual({
      payload: '123-123-123',
      type: 'selectedTerminal/updateSelectedTerminal',
    });
  });

  test('should initiate purchase on clicking Card button', () => {
    // Arrange
    const initiatePurchaseTransaction = jest.spyOn(spiService, 'initiatePurchaseTransaction');
    initiatePurchaseTransaction.mockReturnValue();

    // Arrange
    mockWithRedux(<OrderConfirmation title="title" pathname={PATH_PAY_NOW} currentAmount={500} />, customStore);
    fireEvent.click(screen.getByText(/card/i));

    // Assert
    expect(initiatePurchaseTransaction.mock.calls.length).toBe(1);
  });

  test('should initiate moto purchase on clicking Moto button', () => {
    const initiateMotoPurchaseTransaction = jest.spyOn(spiService, 'initiateMotoPurchaseTransaction');
    initiateMotoPurchaseTransaction.mockReturnValue();

    // Arrange
    mockWithRedux(<OrderConfirmation title="title" pathname={PATH_PAY_NOW} currentAmount={500} />, customStore);
    fireEvent.click(screen.getByText(/moto/i));

    // Assert
    expect(initiateMotoPurchaseTransaction.mock.calls.length).toBe(1);
  });

  test('should cancel moto purchase on clicking Cancel button', () => {
    // Arrange
    const initiateMotoPurchaseTransaction = jest.spyOn(spiService, 'initiateMotoPurchaseTransaction');
    initiateMotoPurchaseTransaction.mockReturnValue();

    const customStore2 = (mockTerminals: Any) => ({
      ...store,
      getState: () => ({
        common: { showFlowPanel: false, acquireConfirmPairingFlow: false },
        terminals: mockTerminals,
        products: {
          keypadAmount: 0,
          surchargeAmount: 100,
          tipAmount: 0,
          cashoutAmount: 100,
          products: [],
        },
        selectedTerminal: { selectedTerminalId: mockTerminalInstanceId },
      }),
    });
    const mockTerminals = {
      [mockTerminalInstanceId]: {
        ...pairedMockTerminals[mockTerminalInstanceId],
        receipt: mockReceiptResponse,
        txFlow: {
          ...mockRefundTxFlow,
          finished: false,
          success: TxFlowState.Unknown,
        },
      },
    };

    const spiCancelTransaction = jest.spyOn(spiService, 'spiCancelTransaction');
    spiCancelTransaction.mockReturnValue();

    mockWithRedux(
      <OrderConfirmation title="title" pathname={PATH_PAY_NOW} currentAmount={500} />,
      customStore2(mockTerminals)
    );

    // Act
    fireEvent.click(screen.getByText(/moto/i));

    fireEvent.click(screen.getByText(/cancel/i));

    // Assert
    expect(spiCancelTransaction.mock.calls.length).toBe(1);
  });

  test('should disable Moto button when tip amount is larger than 0', () => {
    // Arrange
    const initiateRefundTxTransaction = jest.spyOn(spiService, 'initiateRefundTxTransaction');
    initiateRefundTxTransaction.mockReturnValue();

    const customStore2 = (mockTerminals: Any) => ({
      ...store,
      getState: () => ({
        common: { showFlowPanel: false, acquireConfirmPairingFlow: false },
        terminals: mockTerminals,
        products: {
          keypadAmount: 0,
          surchargeAmount: 100,
          tipAmount: 100,
          cashoutAmount: 0,
          products: [],
        },
        selectedTerminal: { selectedTerminalId: mockTerminalInstanceId },
      }),
    });
    const mockTerminals = {
      [mockTerminalInstanceId]: {
        ...pairedMockTerminals[mockTerminalInstanceId],
        receipt: mockReceiptResponse,
        txFlow: {
          ...mockRefundTxFlow,
          finished: false,
          success: TxFlowState.Unknown,
        },
      },
    };

    // Act
    mockWithRedux(
      <OrderConfirmation title="title" pathname={PATH_PAY_NOW} currentAmount={500} />,
      customStore2(mockTerminals)
    );
    const motoButton = screen.getByText(/Moto/i).closest('button');

    // Assert
    expect(motoButton).toHaveAttribute('disabled');
    expect(motoButton).toBeDisabled();
  });

  test('should clearAll products when total amount is changed', () => {
    // Arrange
    const initiateMotoPurchaseTransaction = jest.spyOn(spiService, 'initiateMotoPurchaseTransaction');
    initiateMotoPurchaseTransaction.mockReturnValue();

    const spiCancelTransaction = jest.spyOn(spiService, 'spiCancelTransaction');
    spiCancelTransaction.mockReturnValue();

    // Act
    mockWithRedux(<OrderConfirmation title="title" pathname={PATH_PAY_NOW} currentAmount={600} />, customStore);
    fireEvent.click(screen.getByTestId('orderTotalButton'));
    fireEvent.click(screen.getByText(/5/i));
    fireEvent.click(screen.getByText(/ok/i));

    // Assert
    expect(dispatch.mock.calls[0][0]).toEqual({
      payload: undefined,
      type: 'product/clearProductsOnly',
    });
  });

  test('should close the keypad', async () => {
    // Arrange
    const initiateMotoPurchaseTransaction = jest.spyOn(spiService, 'initiateMotoPurchaseTransaction');
    initiateMotoPurchaseTransaction.mockReturnValue();

    const spiCancelTransaction = jest.spyOn(spiService, 'spiCancelTransaction');
    spiCancelTransaction.mockReturnValue();

    // Act
    mockWithRedux(<OrderConfirmation title="title" pathname={PATH_PAY_NOW} currentAmount={500} />, customStore);
    fireEvent.click(screen.getByTestId('orderTotalButton'));
    fireEvent.click(screen.getByLabelText(/close button/i));

    // Assert
    await waitFor(() => {
      expect(screen.queryByLabelText(/close button/i)).not.toBeInTheDocument();
    });
  });

  test('should show cashout button when pathname is cashout', () => {
    // Arrange
    const initiateCashoutOnlyTxTransaction = jest.spyOn(spiService, 'initiateCashoutOnlyTxTransaction');
    initiateCashoutOnlyTxTransaction.mockReturnValue();

    // Act
    mockWithRedux(<OrderConfirmation title="title" pathname={PATH_CASH_OUT} currentAmount={500} />, customStore);
    const cashoutButton = screen.queryByText(/Cashout/i);
    fireEvent.click(cashoutButton as Element);

    // Assert
    expect(cashoutButton).toBeInTheDocument();
  });

  test('should show refund button when pathname is refund', () => {
    // Arrange
    const initiateRefundTxTransaction = jest.spyOn(spiService, 'initiateRefundTxTransaction');
    initiateRefundTxTransaction.mockReturnValue();

    // Act
    mockWithRedux(<OrderConfirmation title="title" pathname={PATH_REFUND} currentAmount={500} />, customStore);
    const refundButton = screen.queryByText(/Refund/i);
    fireEvent.click(refundButton as Element);

    // Assert
    expect(refundButton).toBeInTheDocument();
  });
});
