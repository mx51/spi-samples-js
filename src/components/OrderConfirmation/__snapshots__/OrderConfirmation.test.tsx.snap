// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <OrderConfirmation /> should match OrderConfirmationPage snapshot test 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-justify-content-xs-center"
  >
    <div
      class="MuiGrid-root makeStyles-gridStyles-3 MuiGrid-item MuiGrid-grid-xs-7"
    >
      <div
        class="MuiBox-root MuiBox-root-20 makeStyles-root-1"
      >
        <div
          class="MuiBox-root MuiBox-root-21"
        >
          <div
            class="MuiBox-root MuiBox-root-22"
          >
            <p
              class="MuiTypography-root makeStyles-payLabel-4 MuiTypography-body1"
            >
              title
            </p>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-orderTotalBtn-6"
            data-testid="orderTotalButton"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <div
                class="MuiPaper-root MuiBox-root MuiBox-root-23 makeStyles-paper-5 MuiPaper-elevation1 MuiPaper-rounded"
              >
                <div
                  class="MuiBox-root MuiBox-root-24 makeStyles-orderTotalInputField-8"
                >
                  $5.00
                </div>
                <div
                  class="MuiBox-root MuiBox-root-25"
                >
                  <span
                    aria-hidden="true"
                    class="material-icons MuiIcon-root"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <p
          class="MuiTypography-root makeStyles-label-9 MuiTypography-body1"
        >
          Select terminal
        </p>
        <hr
          class="MuiDivider-root"
        />
        <div
          aria-label="terminalList"
          class="MuiFormGroup-root makeStyles-radioGroup-7"
          role="radiogroup"
        >
          <div
            class="MuiBox-root MuiBox-root-26"
          >
            <ul
              class="MuiList-root MuiList-padding"
            >
              <li
                class="MuiListItem-root MuiListItem-dense"
              >
                <div
                  class="MuiListItemIcon-root"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-27 MuiRadio-root MuiRadio-colorSecondary makeStyles-radioBtn-12 PrivateSwitchBase-checked-28 Mui-checked MuiIconButton-colorSecondary"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input-30"
                        name="terminal"
                        type="radio"
                        value="123-123-123"
                      />
                      <div
                        class="PrivateRadioButtonIcon-root-31 PrivateRadioButtonIcon-checked-33"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-32"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                          />
                        </svg>
                      </div>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </span>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
                  >
                    test
                  </span>
                  <p
                    class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                  >
                    192.168.1.10 S/N 123-123-123
                  </p>
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
